pkg/logs/sender/sender.go

```diff
@@ -92,7 +105,65 @@ func (s *Sender) send(payload []byte) error {
 	return nil
 }
 
+// func (s *Sender) hasError() bool {
+// 	s.Lock()
+// 	defer s.Unlock()
+// 	return s.lastError != nil
+// }
+
 // shouldStopSending returns true if a component should stop sending logs.
 func shouldStopSending(err error) bool {
 	return err == context.Canceled
 }
+
+// SplitSenders splits a single stream of message into 2 equal streams.
+// Acts like an AND gate in that the input will only block if both outputs block.
+// This ensures backpressure is propagated to the input to prevent loss of measages in the pipeline.
+func SplitSenders(inputChan chan *message.Message, main *Sender, backup *Sender) {
+	go func() {
+		mainSenderHasErr := false
+		backupSenderHasErr := false
+
+		for message := range inputChan {
+			select {
+			case mainSenderHasErr = <-main.hasError:
+			default:
+			}
+
+			select {
+			case backupSenderHasErr = <-backup.hasError:
+			default:
+			}
+
+			if !mainSenderHasErr {
+				select {
+				case main.inputChan <- message:
+				case mainSenderHasErr = <-main.hasError:
+				}
+			}
+
+			if !backupSenderHasErr {
+				select {
+				case backup.inputChan <- message:
+				case backupSenderHasErr = <-backup.hasError:
+				}
+			}
+
+			// If both senders are failing, we want to block the pipeline until at least one succeeds
+			for {
+				if mainSenderHasErr && backupSenderHasErr {
+					select {
+					case main.inputChan <- message:
+						mainSenderHasErr = false
```

Comment:
Hmm.. is sending things to a channel adequate assurance that the sender has no error?